File: knockd_rotator_client.py
#!/usr/bin/env python3

import sys
import os
import hashlib
import datetime
import socket
import time
import math

__VERSION__: str = "2.0.0"

# Constants for knock sequence generation
# The sequence length determines how many ports are in the knock sequence
SEQUENCE_LENGTH = int(os.environ.get("KNOCKD_ROTATOR_LENGTH", 10))
assert SEQUENCE_LENGTH >= 5, "KNOCKD_ROTATOR_LENGTH must be at least 5"

# The secret provides additional security against sequence guessing
SECRET = os.environ.get("KNOCKD_ROTATOR_SECRET")
if not SECRET:
    sys.stderr.write("Error: KNOCKD_ROTATOR_SECRET environment variable must be set\n")
    sys.exit(1)
elif len(SECRET) < 10:
    sys.stderr.write(
        "Error: KNOCKD_ROTATOR_SECRET must be at least 10 characters long\n"
    )
    sys.exit(1)

# PROTO_MODULO determines protocol selection:
# - If PROTO_MODULO is 0, always use TCP
# - If PROTO_MODULO > 0, port % PROTO_MODULO even = tcp, odd = udp
PROTO_MODULO = int(os.environ.get("KNOCKD_ROTATOR_PROTO_MODULO", 0))

# Period modulo determines how frequently the sequence changes (in seconds)
# Default: 21600 (6 hours)
PERIOD_MODULO = int(os.environ.get("KNOCKD_ROTATOR_PERIOD_MODULO", 21600))
if "KNOCKD_ROTATOR_PERIOD_MODULO" not in os.environ:
    sys.stderr.write(
        "Warning: KNOCKD_ROTATOR_PERIOD_MODULO not set, using default of 21600 (6 hours)\n"
    )

# Port list for sequence generation
# Default: 2000-65536 range to avoid requiring elevated privileges
port_str = os.environ.get("KNOCKD_ROTATOR_PORTS", "2000-65536")
PORTS = []
for part in port_str.split(","):
    part = part.strip()
    if "-" in part:
        start, end = part.split("-", 1)
        # Include both ends of the range
        PORTS.extend(list(range(int(start), int(end) + 1)))
    else:
        PORTS.append(int(part))

# Ensure no duplicates
if len(PORTS) != len(set(PORTS)):
    sys.stderr.write("Error: Duplicate ports in KNOCKD_ROTATOR_PORTS are not allowed\n")
    sys.exit(1)

# Ensure at least 2 ports
assert len(PORTS) >= 2, "KNOCKD_ROTATOR_PORTS must contain at least 2 ports"

# Warn if 3 or fewer ports
if len(PORTS) <= 3:
    sys.stderr.write(
        "Warning: Only {} ports available. This significantly reduces security.\n".format(
            len(PORTS)
        )
    )

# Sort ports for consistency
PORTS.sort()

# Minimum entropy required (in bits) for a secure knock sequence
# 40 bits is roughly equivalent to a 12-character alphanumeric password
MIN_ENTROPY_BITS = 40


def calculate_knock_entropy(sequence_length: int, port_count: int) -> float:
    """
    Calculate the entropy of a knock sequence in bits.

    Args:
        sequence_length: Length of the knock sequence
        port_count: Number of possible ports to choose from

    Returns:
        float: Entropy in bits (sequence_length * log2(port_count))
    """
    if port_count <= 1:
        return 0
    return sequence_length * math.log2(port_count)


def calculate_shared_seed(offset: int = 0) -> int:
    """
    Calculate the shared seed based on the current time period with an optional offset.

    Args:
        offset: Integer offset to shift the period (negative for past periods, positive for future)
                Default is 0 (current period)

    Returns:
        int: The calculated seed value based on current UTC time with the applied offset
    """
    current_timestamp = int(datetime.datetime.now(datetime.timezone.utc).timestamp())
    # Calculate the beginning of the current period
    period_start = ((current_timestamp // PERIOD_MODULO) + offset) * PERIOD_MODULO
    assert period_start
    assert len(str(period_start)) > 5, f"Suspicious period start value: {period_start}"
    return period_start


# Keep this for compatibility reasons
__SHARED_SEED__ = calculate_shared_seed()


def generate_knock_sequence(service_name: str, offset: int = 0) -> str:
    """
    Generate a formatted knock sequence for a specific service.

    This function combines port generation and formatting into a single call.
    The sequence is based on the current time period and the service name.
    Protocol is determined by port number: even ports use tcp, odd ports use udp.

    If the provided service_name doesn't end with "_ROTATOR", it will
    automatically be appended to ensure compatibility with knock_rotator_server.py.

    Args:
        service_name: The name of the service to generate a sequence for
        offset: Integer offset to shift the time period (default: 0)

    Returns:
        Formatted sequence string (e.g. "1234:tcp 5678:udp 9012:tcp")

    Raises:
        ValueError: If the configuration provides insufficient entropy
    """
    # Calculate and check entropy before generating sequence
    entropy_bits = calculate_knock_entropy(SEQUENCE_LENGTH, len(PORTS))
    if entropy_bits < MIN_ENTROPY_BITS:
        error_msg = (
            f"Insufficient entropy: {entropy_bits:.2f} bits. "
            f"Required minimum: {MIN_ENTROPY_BITS} bits. "
            f"Increase SEQUENCE_LENGTH or add more ports."
        )
        sys.stderr.write(f"Error: {error_msg}\n")
        raise ValueError(error_msg)

    # Ensure service_name ends with _ROTATOR
    if not service_name.endswith("_ROTATOR"):
        service_name = f"{service_name}_ROTATOR"

    # Calculate seed with the provided offset
    current_seed = calculate_shared_seed(offset)

    # Create the seed for this specific service
    section_seed = f"{current_seed}{service_name}{SECRET}"

    # Generate the ports
    ports = []
    for i in range(1, SEQUENCE_LENGTH + 1):
        # Generate a hash from seed+iteration
        hash_input = f"{section_seed}{i}".encode("utf-8")
        hash_value = hashlib.sha256(hash_input).hexdigest()

        # Take first 8 hex chars and convert to decimal
        decimal = int(hash_value[:8], 16)

        # Select a port from our port list
        port_index = decimal % len(PORTS)
        port = PORTS[port_index]
        assert port in PORTS, f"Port {port} not in allowed port list"
        ports.append(port)

    # Format the sequence with protocol determination
    # Use original format: PORT:tcp or PORT:udp
    if PROTO_MODULO == 0:
        # When PROTO_MODULO is 0, always use TCP
        sequence = " ".join([f"{port}:tcp" for port in ports])
    else:
        # Otherwise use modulo to determine protocol
        sequence = " ".join(
            [f"{port}:{'tcp' if port % PROTO_MODULO == 0 else 'udp'}" for port in ports]
        )

    # Validate the sequence meets our requirements
    for port_proto in sequence.split():
        port_str, proto = port_proto.split(":")
        port = int(port_str)
        assert 2000 <= port <= 65535, f"Port {port} is outside valid range (2000-65535)"
        assert proto in [
            "tcp",
            "udp",
        ], f"Protocol {proto} is not valid (must be tcp or udp)"

    return sequence


def knock_ports(host: str, sequence: str):
    """
    Perform port knocking on a target host using the generated sequence.

    Args:
        host: The target host to knock on
        sequence: Space-separated list of port:protocol pairs
    """
    # Parse the sequence into port/protocol pairs
    for port_proto in sequence.split():
        port, proto = port_proto.split(":")
        port = int(port)

        if proto.lower() == "udp":
            print(f"Knocking UDP port {port}...")
            # For UDP, we create a UDP socket and try to connect
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            try:
                sock.sendto(b"", (host, port))
            finally:
                sock.close()
        else:  # TCP
            print(f"Knocking TCP port {port}...")
            # For TCP, we create a TCP socket and try to connect
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(1)  # Set timeout to 1 second
            try:
                sock.connect((host, port))
            except (socket.timeout, socket.error):
                # Expected to fail if port is closed, which is normal
                pass
            finally:
                sock.close()

        # Small delay between knocks
        time.sleep(0.1)


def main():
    """Main function to parse command line arguments and either generate a sequence or perform port knocking."""
    import argparse

    # Calculate and display entropy information
    entropy_bits = calculate_knock_entropy(SEQUENCE_LENGTH, len(PORTS))
    print(f"Knock sequence entropy: {entropy_bits:.2f} bits")
    if entropy_bits < MIN_ENTROPY_BITS:
        print(
            f"WARNING: Current configuration has low entropy ({entropy_bits:.2f} bits)"
        )
        print(f"Minimum recommended entropy: {MIN_ENTROPY_BITS} bits")

    # parse 'gene' as 'generate' etc
    argv = sys.argv
    if "generate" not in argv and "knock" not in argv:
        for ia, a in enumerate(sys.argv):
            if "generate".startswith(a):
                sys.argv[ia] = "generate"
                break
            elif "knock".startswith(a):
                sys.argv[ia] = "knock"
                break

    parser = argparse.ArgumentParser(
        description="Port knock sequence generator and client"
    )
    subparsers = parser.add_subparsers(dest="mode", help="Operation mode")

    # Generate mode
    gen_parser = subparsers.add_parser("generate", help="Generate a knock sequence")
    gen_parser.add_argument(
        "service_name", help="Name of the service to generate sequence for"
    )
    gen_parser.add_argument(
        "--offset",
        type=int,
        default=0,
        help="Time period offset (negative for past, positive for future)",
    )

    # Knock mode
    knock_parser = subparsers.add_parser("knock", help="Perform port knocking")
    knock_parser.add_argument("host", help="Target host to knock on")
    knock_parser.add_argument(
        "service_name", help="Name of the service to generate sequence for"
    )
    knock_parser.add_argument(
        "--offset",
        type=int,
        default=0,
        help="Time period offset (negative for past, positive for future)",
    )

    args = parser.parse_args()

    if not args.mode:
        parser.print_help()
        sys.exit(1)

    if args.mode == "generate":  # Generate mode
        print(generate_knock_sequence(args.service_name, args.offset))

    elif args.mode == "knock":  # Knock mode
        sequence = generate_knock_sequence(args.service_name, args.offset)
        knock_ports(args.host, sequence)

    else:
        print(f"Unknown mode: {args.mode}")
        parser.print_help()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: KnockdRotatorClient.kt
import java.security.MessageDigest
import java.time.Instant
import java.time.ZoneOffset
import kotlin.math.log2
import kotlin.system.exitProcess

/**
 * Object responsible for generating knock sequences compatible with knockd-rotator.
 *
 * This object encapsulates the logic for generating time-based, secret-dependent
 * port knocking sequences. It can be configured via parameters passed to its
 * generation function, making it suitable for embedding in applications like Android apps.
 */
object KnockdRotatorGenerator {

    // Default configuration values (matching the Python script)
    const val DEFAULT_SEQUENCE_LENGTH = 10
    const val DEFAULT_PROTO_MODULO = 0 // 0 means always TCP
    const val DEFAULT_PERIOD_MODULO = 21600 // 6 hours in seconds
    const val DEFAULT_PORTS_STRING = "2000-65536"
    const val MIN_ENTROPY_BITS = 40.0

    /**
     * Parses a port string (e.g., "2000-65535,8080") into a sorted list of unique integers.
     *
     * @param portStr The string representation of ports.
     * @return A sorted list of unique port numbers.
     * @throws IllegalArgumentException if the format is invalid, ports are duplicated,
     *         or fewer than 2 ports are specified.
     */
    fun parsePorts(portStr: String): List<Int> {
        val ports = mutableSetOf<Int>()
        try {
            portStr.split(",").forEach { part ->
                val trimmedPart = part.trim()
                if (trimmedPart.contains("-")) {
                    val range = trimmedPart.split("-", limit = 2)
                    val start = range[0].toInt()
                    val end = range[1].toInt()
                    if (start > end) throw IllegalArgumentException("Invalid range: $start > $end")
                    (start..end).forEach { ports.add(it) }
                } else if (trimmedPart.isNotEmpty()) {
                    ports.add(trimmedPart.toInt())
                }
            }
        } catch (e: NumberFormatException) {
            throw IllegalArgumentException("Invalid port format in '$portStr'. ${e.message}", e)
        }

        if (ports.size < 2) {
            throw IllegalArgumentException("KNOCKD_ROTATOR_PORTS must contain at least 2 unique ports (found ${ports.size})")
        }

        // Warn if 3 or fewer ports (can be handled by the caller if needed)
        // if (ports.size <= 3) {
        //     System.err.println("Warning: Only ${ports.size} ports available. This significantly reduces security.")
        // }

        return ports.toList().sorted()
    }

    /**
     * Calculates the entropy of a knock sequence in bits.
     *
     * Entropy = sequence_length * log2(port_count)
     *
     * @param sequenceLength Length of the knock sequence.
     * @param portCount Number of possible ports to choose from.
     * @return Entropy in bits. Returns 0.0 if portCount <= 1.
     */
    fun calculateKnockEntropy(sequenceLength: Int, portCount: Int): Double {
        if (portCount <= 1) {
            return 0.0
        }
        return sequenceLength * log2(portCount.toDouble())
    }

    /**
     * Calculates the shared seed based on the current time period with an optional offset.
     * The seed is derived from the start of the current UTC time period defined by periodModulo.
     *
     * @param periodModulo The duration of each period in seconds.
     * @param offset Integer offset to shift the period (negative for past, positive for future). Default is 0.
     * @return The calculated seed value (Unix timestamp of the period start).
     * @throws IllegalStateException if the calculated period start is zero.
     */
    fun calculateSharedSeed(periodModulo: Int, offset: Int = 0): Long {
        // Get current UTC timestamp in seconds
        val currentTimestamp = Instant.now().epochSecond
        // Calculate the start of the relevant period
        val periodStart = ((currentTimestamp / periodModulo) + offset) * periodModulo

        // Basic sanity checks matching Python/Zsh scripts
        if (periodStart == 0L) {
            throw IllegalStateException("Calculated period start is zero. Check periodModulo ($periodModulo) and system time.")
        }
        if (periodStart.toString().length <= 5) {
            // Log warning instead of throwing error for potentially valid short timestamps
            System.err.println("Warning: Suspiciously short period start value: $periodStart")
        }
        return periodStart
    }

    /**
     * Generates a formatted knock sequence for a specific service based on configuration.
     *
     * This is the core logic, designed to be easily called from other Kotlin/Java code.
     *
     * @param serviceName The name of the service (e.g., "ssh"). "_ROTATOR" will be appended if missing.
     * @param offset Time period offset (default: 0).
     * @param sequenceLength The number of ports in the sequence.
     * @param secret The shared secret string (must be >= 10 chars).
     * @param protoModulo Determines protocol (0=always TCP, >0=port%modulo even=TCP, odd=UDP).
     * @param periodModulo How often the sequence changes (seconds).
     * @param ports The list of available ports to choose from (must have >= 2).
     * @return Formatted sequence string (e.g., "1234:tcp 5678:udp 9012:tcp").
     * @throws IllegalArgumentException if configuration is invalid (e.g., short secret, insufficient ports, low entropy).
     */
    @Throws(IllegalArgumentException::class)
    fun generateKnockSequence(
        serviceName: String,
        offset: Int = 0,
        sequenceLength: Int,
        secret: String,
        protoModulo: Int,
        periodModulo: Int,
        ports: List<Int>
    ): String {
        // --- Input Validations ---
        if (sequenceLength < 5) {
            throw IllegalArgumentException("Sequence length must be at least 5.")
        }
        if (secret.length < 10) {
            throw IllegalArgumentException("Secret must be at least 10 characters long.")
        }
        if (ports.size < 2) {
            // This should ideally be caught by parsePorts, but double-check
            throw IllegalArgumentException("Must have at least 2 ports available.")
        }
        // Validate port range (basic check, assumes parsePorts did thorough validation)
        ports.forEach {
             if (it < 1 || it > 65535) throw IllegalArgumentException("Port $it is outside valid range (1-65535)")
        }


        // --- Entropy Check ---
        val entropyBits = calculateKnockEntropy(sequenceLength, ports.size)
        if (entropyBits < MIN_ENTROPY_BITS) {
            val errorMsg = String.format(
                "Insufficient entropy: %.2f bits. Required minimum: %.0f bits. Increase sequence length or add more ports.",
                entropyBits, MIN_ENTROPY_BITS
            )
            // Log warning, but throw exception to match Python's behavior on insufficient entropy
             System.err.println("Warning: $errorMsg")
            throw IllegalArgumentException(errorMsg)
        }

        // --- Sequence Generation ---

        // Ensure service_name ends with _ROTATOR
        val actualServiceName = if (serviceName.endsWith("_ROTATOR")) serviceName else "${serviceName}_ROTATOR"

        // Calculate seed for the target period
        val currentSeed = calculateSharedSeed(periodModulo, offset)

        // Create the service-specific seed string
        val sectionSeed = "$currentSeed$actualServiceName$secret"

        // Prepare SHA-256 hasher
        val digest = MessageDigest.getInstance("SHA-256")

        val generatedPorts = mutableListOf<Int>()
        for (i in 1..sequenceLength) {
            // Generate hash from seed + iteration number
            val hashInput = "$sectionSeed$i".toByteArray(Charsets.UTF_8)
            val hashBytes = digest.digest(hashInput)

            // Take first 4 bytes (8 hex chars) and convert to Long, then Int
            // Need to handle potential sign extension when converting bytes to Int/Long
            var decimalValue: Long = 0
            for (j in 0..3) {
                // Treat byte as unsigned by masking with 0xFF before shifting
                decimalValue = (decimalValue shl 8) or (hashBytes[j].toLong() and 0xFF)
            }

            // Select a port from the list using modulo
            // Use toInt() safely as the index won't exceed list size which is Int range
            // Use absolute value of modulo result in case decimalValue is negative after conversion
            // (though with the unsigned conversion above, it shouldn't be)
            val portIndex = (decimalValue % ports.size).toInt().let { if (it < 0) it + ports.size else it }
            val port = ports[portIndex]
            generatedPorts.add(port)
        }

        // Format the sequence with protocol determination
        val sequenceParts = generatedPorts.map { port ->
            val protocol = if (protoModulo == 0) {
                "tcp"
            } else {
                if (port % protoModulo == 0) "tcp" else "udp"
            }
            "$port:$protocol"
        }

        return sequenceParts.joinToString(" ")
    }
}

/**
 * Main function for command-line execution to generate a knock sequence.
 * Reads configuration from environment variables and arguments.
 * This allows verifying the Kotlin implementation against the Python/Zsh scripts.
 *
 * Usage:
 *   kotlin KnockdRotatorClient.kt generate <service_name> [--offset N]
 *
 * Environment Variables:
 *   KNOCKD_ROTATOR_LENGTH (optional, default 10)
 *   KNOCKD_ROTATOR_SECRET (required, min 10 chars)
 *   KNOCKD_ROTATOR_PROTO_MODULO (optional, default 0)
 *   KNOCKD_ROTATOR_PERIOD_MODULO (optional, default 21600)
 *   KNOCKD_ROTATOR_PORTS (optional, default "2000-65536")
 */
fun main(args: Array<String>) {
    // --- Configuration from Environment Variables ---
    val sequenceLength = System.getenv("KNOCKD_ROTATOR_LENGTH")?.toIntOrNull()
        ?: KnockdRotatorGenerator.DEFAULT_SEQUENCE_LENGTH
    val secret = System.getenv("KNOCKD_ROTATOR_SECRET")
    val protoModulo = System.getenv("KNOCKD_ROTATOR_PROTO_MODULO")?.toIntOrNull()
        ?: KnockdRotatorGenerator.DEFAULT_PROTO_MODULO
    val periodModulo = System.getenv("KNOCKD_ROTATOR_PERIOD_MODULO")?.toIntOrNull()
        ?: KnockdRotatorGenerator.DEFAULT_PERIOD_MODULO
    val portStr = System.getenv("KNOCKD_ROTATOR_PORTS")
        ?: KnockdRotatorGenerator.DEFAULT_PORTS_STRING

    // --- Basic Environment Variable Validation ---
    if (secret == null) {
        System.err.println("Error: KNOCKD_ROTATOR_SECRET environment variable must be set")
        exitProcess(1)
    }
     if (secret.length < 10) {
         System.err.println("Error: KNOCKD_ROTATOR_SECRET must be at least 10 characters long")
         exitProcess(1)
     }
    if (sequenceLength < 5) {
        System.err.println("Error: KNOCKD_ROTATOR_LENGTH must be at least 5")
        exitProcess(1)
    }
    // Check if default period modulo is used without explicit setting
    if (System.getenv("KNOCKD_ROTATOR_PERIOD_MODULO") == null && periodModulo == KnockdRotatorGenerator.DEFAULT_PERIOD_MODULO) {
         System.err.println("Warning: KNOCKD_ROTATOR_PERIOD_MODULO not set, using default of ${KnockdRotatorGenerator.DEFAULT_PERIOD_MODULO} (${KnockdRotatorGenerator.DEFAULT_PERIOD_MODULO / 3600} hours)")
    }


    // --- Parse Ports ---
    val ports: List<Int>
    try {
        ports = KnockdRotatorGenerator.parsePorts(portStr)
    } catch (e: IllegalArgumentException) {
        System.err.println("Error parsing KNOCKD_ROTATOR_PORTS: ${e.message}")
        exitProcess(1)
    }
     // Warn if few ports
     if (ports.size <= 3) {
         System.err.println("Warning: Only ${ports.size} ports available. This significantly reduces security.")
     }


    // --- Argument Parsing (Simple) ---
    var serviceName: String? = null
    var offset = 0
    var mode: String? = null

    val argList = args.toMutableList()

    // Very basic argument parsing for 'generate' mode and '--offset'
    if (argList.isNotEmpty() && (argList[0] == "generate" || argList[0] == "gen")) {
        mode = "generate"
        argList.removeAt(0) // Consume mode

        val offsetIndex = argList.indexOf("--offset")
        if (offsetIndex != -1 && offsetIndex + 1 < argList.size) {
            try {
                offset = argList[offsetIndex + 1].toInt()
                // Remove offset flag and value
                argList.removeAt(offsetIndex)
                argList.removeAt(offsetIndex)
            } catch (e: NumberFormatException) {
                System.err.println("Error: Invalid value for --offset: ${argList[offsetIndex + 1]}")
                exitProcess(1)
            }
        }

        if (argList.size == 1) {
            serviceName = argList[0]
        }
    }

    if (mode != "generate" || serviceName == null) {
        System.err.println("Usage: kotlin KnockdRotatorClient.kt generate <service_name> [--offset N]")
        System.err.println("  (Reads configuration from KNOCKD_ROTATOR_* environment variables)")
        exitProcess(1)
    }

    // --- Calculate and Display Entropy ---
    val entropyBits = KnockdRotatorGenerator.calculateKnockEntropy(sequenceLength, ports.size)
    println(String.format("Knock sequence entropy: %.2f bits", entropyBits))
    if (entropyBits < KnockdRotatorGenerator.MIN_ENTROPY_BITS) {
        System.err.println(String.format("WARNING: Current configuration has low entropy (%.2f bits)", entropyBits))
        System.err.println(String.format("Minimum recommended entropy: %.0f bits", KnockdRotatorGenerator.MIN_ENTROPY_BITS))
        // Note: Generation will fail later if entropy is too low, but we warn here too.
    }


    // --- Generate and Print Sequence ---
    try {
        val sequence = KnockdRotatorGenerator.generateKnockSequence(
            serviceName = serviceName,
            offset = offset,
            sequenceLength = sequenceLength,
            secret = secret, // Already checked for null
            protoModulo = protoModulo,
            periodModulo = periodModulo,
            ports = ports
        )
        println(sequence) // Print the generated sequence to standard output
    } catch (e: IllegalArgumentException) {
        System.err.println("Error generating sequence: ${e.message}")
        exitProcess(1)
    } catch (e: IllegalStateException) {
        System.err.println("Error calculating seed: ${e.message}")
        exitProcess(1)
    } catch (e: Exception) {
        System.err.println("An unexpected error occurred: ${e.message}")
        e.printStackTrace() // Print stack trace for unexpected errors
        exitProcess(1)
    }
}
